uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/WorkPatternCalendar.php

```diff
@@ -0,0 +1,316 @@
+   * which is a list of all the Work Patterns effective for this calendar's
+   * contact, during this calendar's Absence Period, together with information
+   * of during which dates this pattern was active during that period.
+   *
+   * This method take into account the contact's contracts during the period,
+   * and the returned dates are adjusted to match the contracts dates. That is,
+   * even if a contact has an effective work pattern for a given date, the
+   * method won't include that pattern unless the contact also has a contract
+   * during that date.
+   *
+   * @return array
+   *   An array of Work Patterns, organized according to the dates of the
+   *   contacts contracts during the Absence Period. Each entry has:
+   *   - pattern_id: The ID of the Work Pattern
+   *   - effective_date: The date this work pattern became active for one
+   *     specific contract. @see calculateWorkPatternEffectiveDateForContract
+   *   - period_start_date: The start date for this work pattern on the period
+   *     @see calculateWorkPatternPeriodStartDate
+   *   - period_end_date: The end date for this work pattern on the period
+   *     @see calculateWorkPatternPeriodEndDate
+   *
+   *   Given that gaps between contracts might exist and a single work pattern
+   *   might cover more than one contract, the returned array might include more
+   *   than one entry for the same work pattern (one for each contract covered
+   *   by that work pattern), but with different dates.
+   */
+  private function getWorkPatternsPeriods() {
+    $workPatterns = [];
+    $contracts = $this->getContractsWithAdjustedDatesForPeriod();
+
+    foreach($contracts as $contract) {
+      $contractStartDate = new DateTime($contract['period_start_date']);
+      $contractEndDate = new DateTime($contract['period_end_date']);
+      $contractOriginalStartDate = new DateTime($contract['original_start_date']);
+
+      $contactWorkPatterns = ContactWorkPattern::getAllForPeriod(
+        $this->contactID,
+        $contractStartDate,
+        $contractEndDate
+      );
+
+      // If there's no work pattern for this contract, we use the default one
+      // for its whole period
+      if(empty($contactWorkPatterns)) {
+        $workPattern = WorkPattern::getDefault();
+        $workPatterns[] = [
+          'pattern_id' => (int)$workPattern->id,
+          'effective_date' => $contractOriginalStartDate,
+          'period_start_date' => $contractStartDate,
+          'period_end_date' => $contractEndDate
+        ];
+
+        continue;
+      }
+
+      // If the first returned work pattern starts being effective after the
+      // contract start date, it means that, for some time, that contract's
+      // contact didn't have any work pattern assigned. In this case, we have to
+      // use the default work pattern for this period
+      $firstWorkPattern = reset($contactWorkPatterns);
```

Comment:
This is not possible. When a new work pattern is assigned to a contact the the effective_end_date for the current one is automatically set to new pattern effective_date - 1 day